rfc-013-use-docker-to-deploy-and-run-applications-in-containers.md

```diff
@@ -0,0 +1,89 @@
+is continuously growing.
+
+We currently host our work using numerous services and platforms:
+
+- Govpress
+- AWS
+- Azure
+- Heroku
+- Dalmatian
+- Government Platform as a Service (GPaaS)
+
+We cannot reduce the amount of services we use as doing so would reduce the
+flexibility we offer our clients. Clients sometimes require we use their tooling
+and platforms ensuring all their services can be managed centrally. This makes
+their services easier to operate as they move into live. It also makes it
+easier for clients to transition away from dxw.
+
+Most applications are already running with containers but some are not.
+
+There are at least 2 category of problem. The missed advantages of not
+containerising our applications and the problems that come from being
+inconsistent in our approach.
+
+There is a common set of problems for software teams that containerisation can
+help us with:
+
+- managing dependencies
+- updating software
+- debugging issues
+- configuring host machines
+- testing and releasing code
+- rolling back
+
+The problems of being inconsistent include:
+
+- we spend more time at the start of projects debating what tool to choose by
+  default
+- people have to remember the nuances involved in operating each service and
+  attempt to document the variations. These variations must be kept up to date
+  and clearly referenced. Being consistent in our use of containers may allow us
+  to start making our processes for common tasks consistent too.
+  - CI environments that do not test with containers require extra an process in
+    order to maintain a similar environment to production. Configuring tests
+    with containers removes the risk of getting this wrong, and a need for
+    documentation
+  - the process for accessing live environments differs depending on the
+    infrastructure and whether or not containers are used. If containers are
+    used throughout we can have a shared process for accessing a console and
+    running scripts like migrations
+  - the way secrets are distributed to applications can differ. We could
+    manually place them on a machine, we could add them through Terraform and
+    task definitions, we could use a service like Vault or Parameter Store or
+    we could use Heroku to manage the process for us. If we decide to be
+    consistent with the use of containers, we can then start to look at being
+    more consistent with how we distribute secrets in a common way and develop
+    tooling and documentation to make this easier
+
+## Proposal
+
+We MUST use Docker to run our applications in live environments.
+
+We MUST configure continuous integration to build and test the same Docker
+image that is deployed.
+
+We MAY use Docker in development environments.
+
+If there are compelling reasons not to use Docker on a particular project we
+MAY choose not to containerise, however an ADR MUST be added to the
+project to explain and document why.
```

Comment:
The first and last paragraphs are contradictory. We MUST use Docker but we MAY choose not to?

What is an ADR?

Reply:
You're right. Using "SHOULD use Docker" as the main declaration felt too ambiguous. I will attempt to clarify with more words.

Reply:
ADR stands for Architectural Decision Record. It is right not to assume everyone knows the acronym so I will make sure to add both here.